clickhouse.go

```diff
@@ -286,87 +293,46 @@ func DefaultDialStrategy(ctx context.Context, connID int, opt *Options, dial Dia
 }
 
 func (ch *clickhouse) acquire(ctx context.Context) (conn nativeTransport, err error) {
-	timer := time.NewTimer(ch.opt.DialTimeout)
-	defer timer.Stop()
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	default:
+	if ch.closed.Load() {
+		return nil, ErrConnectionClosed
 	}
+
+	ctx, cancel := context.WithTimeoutCause(ctx, ch.opt.DialTimeout, ErrAcquireConnTimeout)
+	defer cancel()
+
 	select {
-	case <-timer.C:
-		return nil, ErrAcquireConnTimeout
-	case <-ctx.Done():
-		return nil, ctx.Err()
 	case ch.open <- struct{}{}:
+	case <-ctx.Done():
+		return nil, context.Cause(ctx)
 	}
-	select {
-	case <-timer.C:
-		select {
-		case <-ch.open:
-		default:
-		}
-		return nil, ErrAcquireConnTimeout
-	case conn := <-ch.idle:
-		if conn.isBad() {
-			conn.close()
-			if conn, err = ch.dial(ctx); err != nil {
-				select {
-				case <-ch.open:
-				default:
-				}
-				return nil, err
-			}
+
+	conn, err = ch.idle.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if conn != nil {
+		if !conn.isBad() {
+			conn.setReleased(false)
+			conn.debugf("[acquired from pool]")
+			return conn, nil
 		}
-		conn.setReleased(false)
-		conn.debugf("[acquired from pool]")
-		return conn, nil
-	default:
+
+		conn.close()
 	}
+
 	if conn, err = ch.dial(ctx); err != nil {
```

Comment:
should we check for `ctx.Done()` before doing `ch.dial()` may be? Because I notice the passed in context only checked again all the way during `queryHello` to the server.